### PR TITLE
(Ground / leg) work for getting look here links / images and such working with comments

### PR DESCRIPTION
**Pull request by jaredpetker**
@cvializ @andrewconner Some beginning work for getting our current message infrastructure working with comments, I rewrote / picked out things that were needed from Jared's old message formatter as the formatter used to to the parsing and formatting in javascript. Now messages are parsed into an array of items that can be used for display. 

Still needs some work for getting some formatting working better and getting look here's working fully, but at least links, images, emails and such will do things similar to the extension. Want to get this in for people to test out while I work on it more.

### CODE REVIEW

**Comment by andrewconner** [comment]
Oh sweet, will go through

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html [hunk 1]
```diff
@@ -12,7 +12,25 @@
             ng-attr-datetime="{{::comment.sentAt}}" am-time-ago="::comment.sentAt"
             title="{{::comment.sentAt|localTime}}" ng-if="::comment.sentAt"></time>
     </div>
-    <span class="kf-keep-comment-commenter-text" ng-bind="comment.text">thing</span>
+    <div class="kf-keep-comment-commenter-text">
+      <div ng-repeat="part in commentParts" class="kf-keep-comment-inline">
+
+        <div  ng-if="part.type === 'IMAGE'" class="kf-keep-comment-block" ng-class="{'kf-keep-comment-v-space': !$last, 'kf-keep-comment-v-space-top': $index - 1 >= 0 && commentParts[$index - 1].type !== 'IMAGE'}">
+          <img width="100px" height="100px" src="{{part.data.src}}"/>
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:19]: - you must use `ng-src`
- the `width` and `height` attributes are deprecated in favor of CSS styling

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html [hunk 2]
```diff
@@ -12,7 +12,25 @@
...
+        </div>
+
+        <div ng-if="part.type === 'TEXT_PLAIN'" ng-bind="part.data.text" class="kf-keep-comment-inline"></div>
+
+        <div ng-if="part.type === 'EMAIL'" class="kf-keep-comment-inline">
+          <a href="mail-to:{{part.data.text}}" ng-bind="part.data.text" class="kf-keep-comment-inline"></a>
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:25]: `ng-href`

nice use of ng-bind here :grinning:

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html [hunk 3]
```diff
@@ -12,7 +12,25 @@
...
+        </div>
+
+        <div ng-if="part.type === 'LINK'" class="kf-keep-comment-inline">
+          <a  href="{{part.data.link}}"  ng-bind="part.data.text" class="kf-keep-comment-inline"></a>
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:29]: `ng-href`

File: kifi-backend/modules/shoebox/angular/src/common/services/messageFormattingService.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+  .factory('messageFormattingService', ['emojiService',
+    function (emojiService) {
+
+      var entityMap = {
+        '&': '&amp;',
+        '<': '&lt;',
+        '>': '&gt;',
+        '"': '&quot;',
+        "'": '&#39;', // jshint ignore:line
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/services/messageFormattingService.js:13]: prefer escaping the single quote in the string like `'\''` over having jshint ignore the line

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html [hunk 5]
```diff
@@ -12,7 +12,25 @@
             ng-attr-datetime="{{::comment.sentAt}}" am-time-ago="::comment.sentAt"
             title="{{::comment.sentAt|localTime}}" ng-if="::comment.sentAt"></time>
     </div>
-    <span class="kf-keep-comment-commenter-text" ng-bind="comment.text">thing</span>
+    <div class="kf-keep-comment-commenter-text">
+      <div ng-repeat="part in commentParts" class="kf-keep-comment-inline">
+
+        <div  ng-if="part.type === 'IMAGE'" class="kf-keep-comment-block" ng-class="{'kf-keep-comment-v-space': !$last, 'kf-keep-comment-v-space-top': $index - 1 >= 0 && commentParts[$index - 1].type !== 'IMAGE'}">
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:18]: feel free to add newlines between attributes and even inside attributes for readability

e.g. 

``` html
<div
  class=""
  ng-if=""
  ng-class="{
    'kf-keep-comment-v-space': !$last
  }"
></div>
```

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.styl [hunk 6]
```diff
@@ -26,7 +33,19 @@
       word-wrap: break-word
       flex-grow: 1
       color: #3E4955FF
-      margin-top: 2px
+      margin-top: 5px
+
+    .kf-keep-comment-inline
+      display: inline
+
+    .kf-keep-comment-block
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.styl:41]: this class is only applied to a div, and divs are `display: block` by default, so this class is unnecessary

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html [hunk 7]
```diff
@@ -12,7 +12,25 @@
             ng-attr-datetime="{{::comment.sentAt}}" am-time-ago="::comment.sentAt"
             title="{{::comment.sentAt|localTime}}" ng-if="::comment.sentAt"></time>
     </div>
-    <span class="kf-keep-comment-commenter-text" ng-bind="comment.text">thing</span>
+    <div class="kf-keep-comment-commenter-text">
+      <div ng-repeat="part in commentParts" class="kf-keep-comment-inline">
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:16]: technically it's invalid for an inline element to contain a block element. consider changing this CSS class's rule to `display: inline-block`

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.js [hunk 8]
```diff
@@ -2,16 +2,16 @@
...
+    function (messageFormattingService) {
       return {
         restrict: 'A',
         scope: {
           comment: '='
         },
         templateUrl: 'common/directives/comments/comment/comment.tpl.html',
-        link: function () {
-
+        link: function (scope) {
```
Comment (cvializ) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.js:13]: going forward, let's prefer `$scope` over `scope`

**Comment by cvializ** [comment]
noice it's looking good

File: kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html (hunk 1 above)
Comment (jaredpetker) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:19]: sure, mostly had the width/height there when I was initially templating

Comment (jaredpetker) [inline kifi-backend/modules/shoebox/angular/src/common/directives/comments/comment/comment.tpl.html:19]: will also be changing this to the whole <div> background url stuff anyways to have better control over the scaling

File: kifi-backend/modules/shoebox/angular/src/common/services/messageFormattingService.js (hunk 4 above)
Comment (jaredpetker) [inline kifi-backend/modules/shoebox/angular/src/common/services/messageFormattingService.js:13]: these no longer exist anyway, there was escaping I copy / pasted from how the extension did things but it's probably not needed anymore.

**Pull request merged**